File: 125 Valid Palindrome.go
package main
/*
125 Valid Palindrome
https://oj.leetcode.com/problems/valid-palindrome/
*/
import (
	"fmt"
    "bytes"
)
func isPalindromicString(y string) bool {
	var delimiters []uint8 = []uint8{' ',',','.',':',';'}
	var x []uint8 = []uint8(y)

	x = bytes.ToLower(x)
	i,j := 0,len(x)-1
	for i<=j{
		if bytes.IndexByte(delimiters,x[i]) != -1{
			i++
		} else if bytes.IndexByte(delimiters,x[j]) != -1{
			j--
		} else {
			if x[i] != x[j] {
				return false
			} else {
				i++
				j--
			}
		}
	}
    return true
}
func main() {
    ls := "A man, a plan, a canal: Panama"
    fmt.Printf("%v\n",isPalindromicString(ls))
    
    ls = "race a car"
    fmt.Printf("%v\n",isPalindromicString(ls))
}

File: 125 Valid Palindrome.py
'''
125. Valid Palindrome
https://leetcode.com/problems/valid-palindrome/

Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Note: For the purpose of this problem, we define empty string as valid palindrome.

Example 1:

Input: "A man, a plan, a canal: Panama"
Output: true

Example 2:

Input: "race a car"
Output: false
'''
class Solution:
    '''
    Re-trim + string comparison, faster
    '''
    def isPalindrome(self, s: str) -> bool:
        trimmed = re.sub("\W", "", s).lower()
        size = len(trimmed)
        halfSize = (size >> 1)
        if (size & 1) == 1:
            return trimmed[:halfSize] == trimmed[:halfSize:-1]
        else:
            return trimmed[:halfSize] == trimmed[:halfSize-1:-1]
        
class Solution:
    '''
    Two Pointers, slower
    '''
    def isPalindrome(self, s: str) -> bool:
        i = 0
        j = len(s) - 1
        while i < j:
            if not s[i].isalnum():
                i += 1
            elif not s[j].isalnum():
                j -= 1
            else:
                if s[i].lower() != s[j].lower():
                    return False
                i += 1
                j -= 1
        return True
